Fix a number of issues with divs in ValueRangeAnalysis

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100547

This PR:
- Adds `floordiv` and `truncdiv` as they were missing
- Maps `div` to its correct definition (it was being mapped to `floordiv`)
- Simplifies the bounds of `floordiv`
- Fixes some issues with the returned types of `floor` `ceil`
- Adds tests for the previous point

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

lezcano committed:
Fix a number of issues with divs in ValueRangeAnalysis

This PR:
- Adds `floordiv` as it was missing
- Maps `div` to `truediv` (it was being mapped to `floordiv`
- Simplifies the bounds of `floordiv`
- Fixes some issues with the returned types of `floor` `ceil`
- Adds tests for the previous point

[ghstack-poisoned]

peterbell10 commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -307,17 +307,30 @@ def neg(x):
-        # We think of this as floor(a / b)
-        out = ValueRangeAnalysis.truediv(a, b)
-        return ValueRangeAnalysis.floor(out)
+        return ValueRangeAnalysis.truediv(a, b)
```
`div` should be truncation division for integers and true division for floats.

lezcano committed:
Update on "Fix a number of issues with divs in ValueRangeAnalysis"


This PR:
- Adds `floordiv` as it was missing
- Maps `div` to `truediv` (it was being mapped to `floordiv`
- Simplifies the bounds of `floordiv`
- Fixes some issues with the returned types of `floor` `ceil`
- Adds tests for the previous point

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano committed:
Update on "Fix a number of issues with divs in ValueRangeAnalysis"


This PR:
- Adds `floordiv` and `truncdiv` as they were missing
- Maps `div` to its correct definition (it was being mapped to `floordiv`)
- Simplifies the bounds of `floordiv`
- Fixes some issues with the returned types of `floor` `ceil`
- Adds tests for the previous point

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented on test/test_sympy_utils.py:
```diff
@@ -207,12 +222,17 @@ def test_binary_ref_range(self, fn):
 class TestSympyInterp(TestCase):
     @parametrize("fn", UNARY_OPS + BINARY_OPS + UNARY_BOOL_OPS + BINARY_BOOL_OPS + COMPARE_OPS)
     def test_interp(self, fn):
+        # SymPy does not implement truncation for Expressions
+        if fn in ("div", "truncdiv"):
+            return
```
This is a bit of a pain. See https://github.com/sympy/sympy/issues/14438

ezyang commented on test/test_sympy_utils.py:
```diff
@@ -100,32 +100,47 @@ def generate_range(vals):
+                    continue
+                ref_r = getattr(ReferenceAnalysis, fn)(a, b)
+
+                # These min/max ops have a dtype dependent that's value-dependent on Python
```
As I've mentioned to @nkaretnikov, the correct fix is to NOT use Python semantics, but instead have an alternate min/max implementation that always promotes based on a static promotion behavior.

lezcano replied:
Done. Alas the current implementation doesn't work for `sympy.Expr`s, but this part of the code is not executed for dynamic shapes, so that's fine. I'll as Aaron how to fix this for Expressions.

ezyang commented on torch/utils/_sympy/reference.py:
```diff
@@ -76,8 +76,25 @@ def truediv(a, b):
+        if a.is_integer:
+            return ReferenceAnalysis.truncdiv(a, b)
+        else:
+            return ReferenceAnalysis.truediv(a, b)
```
nyeh, I don't understand. If b isn't an integer, you'll still end up with truediv no?

lezcano replied:
Actually, now that I think about it, it's just when `a` is not an integral that dispatches to `div`, so `div` should be just `truediv` when on Inductor IR. cc @peterbell10

ezyang reviewed (approved):
but please check question about truediv/floordiv type selection

lezcano committed:
Update on "Fix a number of issues with divs in ValueRangeAnalysis"


This PR:
- Adds `floordiv` and `truncdiv` as they were missing
- Maps `div` to its correct definition (it was being mapped to `floordiv`)
- Simplifies the bounds of `floordiv`
- Fixes some issues with the returned types of `floor` `ceil`
- Adds tests for the previous point

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Fix a number of issues with divs in ValueRangeAnalysis"


This PR:
- Adds `floordiv` and `truncdiv` as they were missing
- Maps `div` to its correct definition (it was being mapped to `floordiv`)
- Simplifies the bounds of `floordiv`
- Fixes some issues with the returned types of `floor` `ceil`
- Adds tests for the previous point

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
